### PR TITLE
add R.none

### PR DESCRIPTION
**Pull request by josepot**
I keep missing a function named `R.none`, which would be the counterpart of [`R.all`](http://ramdajs.com/docs/#all). It's easy to accomplish the same thing by doing `R.all(R.not(fn))` but that's not very readable.

In some occasions I've found myself doing things like:

    var foo = R.map(parseInt, ['1', '2', 'sfa']);
    var noneAreNaN = R.all(R.not(isNaN));
    if (noneAreNaN(foo)) {
      // Do something
    }

And I would like to be able to do this:

    var foo = R.map(parseInt, ['1', '2', 'sfa']);
    if (R.none(isNaN)(foo)) {
      // Do something
    }

### CODE REVIEW

File: dist/ramda.js [hunk 1]
```diff
@@ -2330,6 +2335,28 @@
...
+     * @sig (a -> Boolean) -> [a] -> Boolean
+     * @param {Function} fn The predicate function.
+     * @param {Array} list The array to consider.
+     * @return {Boolean} `true` if the predicate is not satisfied by every element, `false` otherwise.
+     * @example
+     *
+     *      var xs = R.map(parseInt, ['1', '2', '3', 'foo']);
+     *      xs; //=> [1, 2, 3, NaN]
+     *      R.none(isNaN)(xs); //=> false
+     *      xs.pop();
```
Comment (buzzdecafe) [inline dist/ramda.js]: should we merge this, i don't think we should have the example *mutate* the list. Better to create e.g. `ys = [1, 2, 3]`

File: test/none.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    it('returns false if any element satisfies the predicate', function() {
+        assert.strictEqual(R.none(even, [1, 3, 5, 7, 8, 11]), false);
+    });
+
+    it('returns true for an empty list', function() {
+        assert.strictEqual(R.none(T, []), true);
+    });
+
+    it('short-circuits on first false value', function() {
```
Comment (buzzdecafe) [inline test/none.js]: this seems like an implementation detail to me. It's nice that it has this behavior, but do we really need to test for it?

Comment (josepot) [inline test/none.js]: @buzzdecafe The only thing that I did was adapt [the tests of `R.all`](https://github.com/ramda/ramda/blob/master/test/all.js) for `R.none`, in order to make it consistent with the current code base. But I can remove these tests if you want.

File: dist/ramda.js (hunk 1 above)
Comment (josepot) [inline dist/ramda.js]: :+1:

File: test/none.js (hunk 2 above)
Comment (buzzdecafe) [inline test/none.js]: was wondering about that. I probably should have looked for myself :smile: 

it's probably me who added that test, too

Comment (josepot) [inline test/none.js]: np! :)

**Comment by davidchambers** [comment]
This function seems like a reasonable addition.

File: dist/ramda.js [hunk 3]
```diff
@@ -6354,6 +6385,7 @@
         multiply: multiply,
         nAry: nAry,
         negate: negate,
+        none: none,
```
Comment (davidchambers) [inline dist/ramda.js]: We're no longer updating __dist/ramda.js__ with each commit. Could you rebase your branch and undo the changes to this file?

File: src/internal/_none.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+module.exports = function _none(fn, list) {
+    var idx = -1;
+    while (++idx < list.length) {
+        if (fn(list[idx])) {
+            return false;
+        }
+    }
+    return true;
+};
```
Comment (davidchambers) [inline src/internal/_none.js]: There's no need for this internal function, since it's only used in the definition of `R.none`.

File: src/none.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+var _curry2 = require('./internal/_curry2');
+var _none = require('./internal/_none');
+/**
```
Comment (davidchambers) [inline src/none.js]: Nit: Two blank lines after imports.

File: src/none.js [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ * Returns `true` if no elements of the list match the predicate,
+ * `false` if there are any that do.
```
Comment (davidchambers) [inline src/none.js]: This line could be simplified to:

     * `false` otherwise.

File: src/none.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ * @param {Function} fn The predicate function.
+ * @param {Array} list The array to consider.
+ * @return {Boolean} `true` if the predicate is not satisfied by every element, `false` otherwise.
+ * @example
+ *
+ *      var xs = R.map(parseInt, ['1', '2', '3', 'foo']);
+ *      xs; //=> [1, 2, 3, NaN]
+ *      var ys = [1, 2, 3];
+ *      R.none(isNaN)(xs); //=> false
+ *      R.none(isNaN)(ys); //=> true
```
Comment (davidchambers) [inline src/none.js]: You could provide both arguments at once in these examples.

File: test/none.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        var test = function(n) {count++; return even(n);};
+        var result = R.none(test, [1, 3, 5, 6, 7, 9]);
+        assert(!result);
+        assert.strictEqual(count, 4);
+    });
+
+    it('works with more complex objects', function() {
+        var xs = [{x: 'abcd'}, {x: 'adef'}, {x: 'fghiajk'}];
+        function len3(o) { return o.x.length === 3; }
+        function hasA(o) { return o.x.indexOf('a') > -1; }
```
Comment (davidchambers) [inline test/none.js]: I'd prefer to see `>= 0` in place of `> -1`.

File: test/none.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        assert.strictEqual(R.none(len3, xs), true);
+        assert.strictEqual(R.none(hasA, xs), false);
+    });
+
+    it('is automatically curried', function() {
+        var count = 0;
+        var test = function(n) {count++; return even(n);};
+        assert(R.none(test)([1, 3, 5, 6, 7, 9]) === false);
```
Comment (davidchambers) [inline test/none.js]: Nit: Use `assert.strictEqual` rather than `assert`.

File: test/none.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+    it('returns true for an empty list', function() {
+        assert.strictEqual(R.none(T, []), true);
+    });
+
+    it('short-circuits on first false value', function() {
+        var count = 0;
+        var test = function(n) {count++; return even(n);};
+        var result = R.none(test, [1, 3, 5, 6, 7, 9]);
+        assert(!result);
```
Comment (davidchambers) [inline test/none.js]: Nit: Use `assert.strictEqual` rather than `assert`.

**Comment by josepot** [comment]
@davidchambers :+1: Please let me know if there is anything else that you would like me to change. Thanks!

File: test/none.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        function len3(o) { return o.x.length === 3; }
+        function hasA(o) { return o.x.indexOf('a') >= 0; }
+        assert.strictEqual(R.none(len3, xs), true);
+        assert.strictEqual(R.none(hasA, xs), false);
+    });
+
+    it('is automatically curried', function() {
+        var count = 0;
+        var test = function(n) {count++; return even(n);};
+        assert.strictEqual(R.none(test)([1, 3, 5, 6, 7, 9]), false);
```
Comment (davidchambers) [inline test/none.js]: Since we're not asserting `count`, this could be `R.none(even)` rather than `R.none(test)`.

File: test/none.js (hunk 2 above)
Comment (davidchambers) [inline test/none.js]: I agree with @buzzdecafe that we should remove this test. We should leave the order in which we test the elements unspecified to avoid tying ourselves to a particular implementation.

Comment (CrossEye) [inline test/none.js]: Yes, we should remove it from `R.all` as well.

It seemed to make sense at the time, but I wouldn't want to imagine how we'd try to explain it if we were ever to re-implement in a concurrent alternative on some imaginary multi-threaded version of the language.  So it's definitely in fact an implementation detail.

**Comment by CrossEye** [comment]
I agree that this function makes sense.  Some sugar is worth having.

File: test/none.js (hunk 2 above)
Comment (davidchambers) [inline test/none.js]: #914

File: src/none.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+ * @memberOf R
+ * @category List
+ * @sig (a -> Boolean) -> [a] -> Boolean
+ * @param {Function} fn The predicate function.
+ * @param {Array} list The array to consider.
+ * @return {Boolean} `true` if the predicate is not satisfied by every element, `false` otherwise.
+ * @example
+ *
+ *      R.none(isNaN)([1, 2, 3, NaN]); //=> false
+ *      R.none(isNaN)([1, 2, 3]); //=> true
```
Comment (davidchambers) [inline src/none.js]: I like your simplification of the examples. :)

I think it would read better with the `true` example before the `false` one, but I don't know *why* I think this. Also, it would be clearer to provide both arguments at once.

     *      R.none(isNaN, [1, 2, 3]); //=> true
     *      R.none(isNaN, [1, 2, 3, NaN]); //=> false

Aside from this niggle, the pull request looks great!

File: src/none.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+ */
+module.exports = _curry2(function(fn, list) {
+    var idx = -1;
+    while (++idx < list.length) {
```
Comment (davidchambers) [inline src/none.js]: We should cache `list.length`:

```javascript
var idx = -1;
var len = list.length;
while (++idx < len) {
```

Comment (josepot) [inline src/none.js]: @davidchambers :+1:  np! I've changed it already. However, I don't quite understand why: I was under the impression that accessing the length of an Array had a complexity of O(1), so I don't see how this will make it different. Would you mind explaining what I'm missing here? Thanks!

Comment (davidchambers) [inline src/none.js]: Accessing the `length` property of an array is O(1), but we're doing it N times. If `list` contains 100 items, caching the value of the `length` property will save 99 property accesses. It's a minor optimization really, but we do it elsewhere and it's nice to be consistent.

**Comment by davidchambers** [comment]
:deciduous_tree: Very nice patch, @josepot! I'll merge as soon as Travis has finished running the tests.

**Comment by megawac** [comment]
I feel like this one should be implemented via composition

**Comment by buzzdecafe** [comment]
:+1:  @megawac 

`not(any(f))` 

doesn't seem too bad to me

**Comment by CrossEye** [comment]
I feel about this one the same as I do about `reject(fn)` as sugar for `filter(not(fn))`.  It's certainly not essential, but it makes code more readable and comes at little cost.

**Comment by megawac** [comment]
I'm fine with `R.none` being implemented but I'd rather see it implemented as `R.none = R.not(R.any)`

**Comment by davidchambers** [comment]
> I'm fine with `R.none` being implemented but I'd rather see it implemented as `R.none = R.not(R.any)`

Excellent idea! Could you switch to this implementation, @josepot?

**Comment by josepot** [comment]
@davidchambers I can do that, np! However, performance-wise I think that the current implementation would be slightly better.

In order to do that I will have to move the [`not`](https://github.com/ramda/ramda/blob/master/src/not.js#L21) function into the 'internal' folder, so that I can reuse it there. Is that ok?

**Comment by megawac** [comment]
@josepot it should perform comparably (just one extra call to not) if not better (possible short circuiting from the transducers)

**Comment by davidchambers** [comment]
> In order to do that I will have to move the [`not`](https://github.com/ramda/ramda/blob/master/src/not.js#L21) function into the 'internal' folder, so that I can reuse it there. Is that ok?

Sounds good!

**Comment by josepot** [comment]
@davidchambers @megawac @buzzdecafe done! Please, let me know if that's what you were looking for. Thanks!

**Comment by CrossEye** [comment]
> In order to do that I will have to move the [`not`](https://github.com/ramda/ramda/blob/master/src/not.js#L21) function into the 'internal' folder, so that I can reuse it there. Is that ok?

I think this is a mistaken impression.  While there's nothing wrong with doing this, it's also not essential.  I would actually like to reduce the number of `_internal` functions used.

**Comment by davidchambers** [comment]
LGTM. I'll let @CrossEye merge this, in case he has reservations.

**Comment by CrossEye** [comment]
No reservations, just that observation.
